### PR TITLE
AP_Proximity: Filter land from drivers using downward facing rangefinder

### PR DESCRIPTION
**Pull request by rishabsingh3003**
Copter maintains downward facing rangefinder values (with correction for tilt),  for surface tracking. In this PR I send the value down to proximity library, and project a 3D vector of obstacles that are being detected by the lidars (rotation of vehicle is included in the calculations).
Then I compare the 'Z' of the obstacle vector, and the downward rangefinder value.. if they are close together, we can assume it to be land and filter it out.

Major use case for this is when user tilts forward by a lot, and the lidar picks up ground at low alt's. This is then sent to the avoidance and vehicle starts reacting to it.

### CODE REVIEW

**Comment by rishabsingh3003** [comment]
A demonstration can be seen here: https://youtu.be/qax2VSB1Tzo

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 1]
```diff
@@ -150,30 +150,12 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
...
-    // @Values: 0:Default,1:Upside Down
-    // @User: Standard
-    AP_GROUPINFO("2_ORIENT", 17, AP_Proximity, _orientation[1], 0),
-
-    // @Param: 2_YAW_CORR
-    // @DisplayName: Second Proximity sensor yaw correction
-    // @Description: Second Proximity sensor yaw correction
-    // @Units: deg
-    // @Range: -180 180
+    // @Param{Copter}: _LAND_DETECT
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:153]: How about we renamed this parameter to "_IGN_GND"?  Let's rename "_land" in the methods to "_ground" as well.

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 2]
```diff
@@ -150,30 +150,12 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
...
+    // @DisplayName: Proximity sensor land detection
+    // @Description: Filter out Proximity data that might be detecting land below the vehicle. This needs a valid downward facing rangefinder reading
+    // @Values: 0: Disabled, 1: Enabled
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:156]: Let's remove the space between "0:" and "Disabled".

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 3]
```diff
@@ -150,30 +150,12 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
...
-    AP_GROUPINFO("2_ORIENT", 17, AP_Proximity, _orientation[1], 0),
-
-    // @Param: 2_YAW_CORR
-    // @DisplayName: Second Proximity sensor yaw correction
-    // @Description: Second Proximity sensor yaw correction
-    // @Units: deg
-    // @Range: -180 180
+    // @Param{Copter}: _LAND_DETECT
+    // @DisplayName: Proximity sensor land detection
+    // @Description: Filter out Proximity data that might be detecting land below the vehicle. This needs a valid downward facing rangefinder reading
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:155]: maybe change the description to "Ignore proximity data that is within 0.5m of the ground below the vehicle.  This requires a downward facing rnagefinder".

File: libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp [hunk 4]
```diff
@@ -94,7 +94,7 @@ void AP_Proximity_TeraRangerTower::update_sector_data(int16_t angle_deg, uint16_
 {   
     // Get location on 3-D boundary based on angle to the object
     const AP_Proximity_Boundary_3D::Face face = boundary.get_face(angle_deg);
-    if (distance_cm != 0xffff) {
+    if (distance_cm != 0xffff && !check_obstacle_near_land(angle_deg, distance_cm * 0.0001f)) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp:97]: is there perhaps one too many zeros here?  I think maybe it should be 0.001f which is 1/1000.

File: libraries/AP_Proximity/AP_Proximity_TeraRangerTowerEvo.cpp [hunk 5]
```diff
@@ -149,7 +149,7 @@ void AP_Proximity_TeraRangerTowerEvo::update_sector_data(int16_t angle_deg, uint
     const AP_Proximity_Boundary_3D::Face face = boundary.get_face(angle_deg);
     //check for target too far, target too close and sensor not connected
     const bool valid = distance_cm != 0xffff && distance_cm != 0x0000 && distance_cm != 0x0001;
-    if (valid) {
+    if (valid && !check_obstacle_near_land(angle_deg, distance_cm * 0.0001f)) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_TeraRangerTowerEvo.cpp:152]: I think this may also have one too many zeros.

**Comment by rmackay9** [comment]
I've put a few things to check but in general it seems pretty good, thanks!

**Comment by rmackay9** [comment]
Just for the record, it is possible that this "ignore ground" feature could be moved to the AC_Avoidance library (perhaps into the boundary) if we took into account the vehicle's lean angle within AC_Avoidance.  To try and clarify a bit more, the distances from horizontally facing lidar are not current corrected for the vehicle's lean angle before being mapped to the appropriate "face" in the boundary.  If we did account for the vehicle' lean angle we could then either throw away those readings before adding them to the boundary or the simple avoidance calculations could ignore faces that are likely close to the ground.  This could mean that the distances still appear in the radar view of the GCS (which I think only shows raw distances) but ignored for avoidance purposes.

File: libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp (hunk 4 above)
Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp]: Rookie mistake! Thanks for catching it

**Comment by rishabsingh3003** [comment]
The biggest change made in the latest push, is that I have disabled this feature when Vehicle is disarmed. While the vehicle is within one meter of the ground, this PR will not let any obstacles be saved in the boundary, and thus no obstacles will be sent to GCS. I am sure we are going to receive a lot of  support questions, where people can't see anything in the proximity viewer, when really its just this feature in play. 
The bad part is that, this means OA DB is still going to get data, while the vehicle is disarmed. But we have other "low alt" features for OA DB...

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp [hunk 6]
```diff
@@ -79,6 +82,83 @@ bool AP_Proximity_Backend::ignore_reading(uint16_t angle_deg) const
...
+    if (!frontend._ign_gnd_enable) {
+        return false;
+    }
+    if (!hal.util->get_soft_armed()) {
+        // don't run this feature while vehicle is disarmed, otherwise proximity data will not show up on GCS
+        return false;
+    }
+    // Assume object is yaw and pitch bearing and distance meters away from the vehicle
+    Vector3f object_3D;
+    object_3D.offset_bearing(wrap_180(yaw), wrap_180(-pitch), distance);
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Backend.cpp:133]: I don't think we can use wrap_180 on the pitch.. so I think we should remove it or throw it away if the pitch is >90 or less than -90.

**Comment by magicrub** [comment]
Is this something plane should start using? Might be nice to merge AP_rangefinder into this library? Not this PR though.. just thinking out loud

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 7]
```diff
@@ -473,6 +455,17 @@ bool AP_Proximity::sensor_failed() const
...
 }
 
+// set alt as read from dowward facing rangefinder. Tilt is already adjusted for.
+void AP_Proximity::set_rangefinder_alt(bool use, bool healthy, float alt_cm)
+{
+    if (!valid_instance(primary_instance)) {
+        return;
+    }
+    // get upward distance from backend
+    return drivers[primary_instance]->set_rangefinder_alt(use, healthy, alt_cm);
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:465]: I think we can remove the return because this function is void.

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp [hunk 8]
```diff
@@ -79,6 +82,83 @@ bool AP_Proximity_Backend::ignore_reading(uint16_t angle_deg) const
             }
         }
     }
+
+    // check if obstacle is near land
+    if (check_obstacle_near_ground(angle_deg, distance_m)) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Backend.cpp:87]: might as well shorten this to just, "return check_obstacle_near_ground.."

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp [hunk 9]
```diff
@@ -79,6 +82,83 @@ bool AP_Proximity_Backend::ignore_reading(uint16_t angle_deg) const
...
+        return false;
+    }
+
+    float alt = FLT_MAX;
+    if (!get_rangefinder_alt(alt)) {
+        return false;
+    }
+
+    if (obstacle.z > -0.5f) {
+        // obstacle is at the most 0.5 meters above vehicle
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Backend.cpp:156]: "above vehicle" or "above ground"?

Comment (rmackay9) [reply libraries/AP_Proximity/AP_Proximity_Backend.cpp:156]: Ah, I see.. if it is 0.5m above the vehicle then we assume it is above the ground?

File: libraries/AP_Proximity/AP_Proximity_Backend.h [hunk 10]
```diff
@@ -20,6 +20,9 @@
 #include <AP_Common/Location.h>
 #include "AP_Proximity_Boundary_3D.h"
 
+#define PROXIMITY_GND_DETECT_THRESHOLD 1.0f // set ground detection threshold to be 1 meters
+#define PROXIMITY_ALT_DETECT_TIMEOUT_MS 500 // alt readings should arrive within this much time
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Backend.h:24]: Not too important but I guess this timeout is for the rangefinder?  Maybe we should call it PROXIMITY_RNGFND_TIMEOUT_MS?

File: libraries/AP_Proximity/AP_Proximity_MAV.cpp [hunk 11]
```diff
@@ -226,11 +228,19 @@ void AP_Proximity_MAV::handle_obstacle_distance_3d_msg(const mavlink_message_t &
     Matrix3f body_to_ned;
     const bool database_ready = database_prepare_for_push(current_pos, body_to_ned);
 
-    const Vector3f obstacle(packet.x, packet.y, packet.z * -1.0f);
-    if (obstacle.length() < _distance_min || obstacle.length() > _distance_max || obstacle.is_zero()) {
+    const Vector3f obstacle_FRD(packet.x, packet.y, packet.z);
+    if (obstacle_FRD.length() < _distance_min || obstacle_FRD.length() > _distance_max || obstacle_FRD.is_zero()) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_MAV.cpp:232]: a small optimisation here would be to calculate obstacle_FRD.length just once.  It uses a sqrt so it does take a bit of CPU to calculate it.

File: libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp [hunk 12]
```diff
@@ -94,7 +94,7 @@ void AP_Proximity_TeraRangerTower::update_sector_data(int16_t angle_deg, uint16_
 {   
     // Get location on 3-D boundary based on angle to the object
     const AP_Proximity_Boundary_3D::Face face = boundary.get_face(angle_deg);
-    if (distance_cm != 0xffff) {
+    if (distance_cm != 0xffff && !check_obstacle_near_ground(angle_deg, distance_cm * 0.001f)) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_TeraRangerTower.cpp:97]: sorry to be picky but can we put brackets around, "distance_cm != 0xffff"?  i.e. make it "(distance_cm != 0xffff) && ..."

**Comment by rmackay9** [comment]
Sorry for the trickle feed of requested changes.  I think I've got them all now and this is looking pretty good.

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp (hunk 6 above)
Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity_Backend.cpp]: Done. Fixed this in master at another place also

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp (hunk 9 above)
Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity_Backend.cpp:147]: Yes. Primary logic was that since we are checking for a threshold (which is 1 meter).. so any obstacle while flying below 1 meter height is going to be automatically marked as ground. So in that case, if someone comes from the top of the vehicle.. we should probably not ignore it, which is the purpose of this check. We can remove it if you don't like it

File: libraries/AP_Proximity/AP_Proximity_Backend.h (hunk 10 above)
Comment (rishabsingh3003) [reply libraries/AP_Proximity/AP_Proximity_Backend.h:24]: I left this one as it is because I don't want some new developer to be confused with the forward facing rangefinder driver that is also part of the proximity lib. 
Although the only source of alt readings for the proximity lib is downward facing rangefinder, but who knows, if sometime we change it.. probably with the 3D cameras then this name won't be valid. 

But i'll be happy to rename if you still think its a good idea. :)

**Comment by rmackay9** [comment]
Merged, thanks!
